_posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md

```diff
@@ -0,0 +1,110 @@
+
+## Example: boundaries
+
+A boundary value could be forgotten when writing tests (even following TDD). For example, this piece of production code:
+
+```
+status arg3 = ((from.getParam1() < from.getParam2())? BLACK: WHITE);
+```
+
+If we don’t have a test which considers the same value for `param1` and `param2`, a mutation will survive when applying <a href="http://pitest.org/quickstart/mutators/#CONDITIONALS_BOUNDARY" target="_blank">_Conditional Boundary Mutator_</a>:
+
+<center>
+<img src="{{site.baseurl}}/assets/img/custom/blog/2017-08-29-coverage/survived-mutation.png" alt="Survived mutation" class="img img-responsive"/>
+</center>
+<br/>
+
+PIT report shows the affected line:
+
+<center>
+<img src="{{site.baseurl}}/assets/img/custom/blog/2017-08-29-coverage/pit-report-boundaries.png" alt="PIT Report" class="img img-responsive"/>
+</center>
+<br/>
+
+## Example: equals and hashCode methods
+
+I try to avoid having logic in _production_ code which is only used from _test_ code.
+
+It’s common to find `equals` and `hashCode` methods in Java which are only used in _verifications_ or _assertions_. It’s easy to generate the code of these methods automatically with an IDE such as _IntelliJ IDEA_, but at the same time, it’s easy to have outdated code if we don’t remember to regenerate them when changing the involved class (or we don’t receive an alert about this fact).
+
+For example, a property is added to a class without updating `equals` and `hashCode` methods, so PIT statistics results in:
+
+<center>
+<img src="{{site.baseurl}}/assets/img/custom/blog/2017-08-29-coverage/pit-statistics.png" alt="PIT Statistics" class="img img-responsive"/>
+</center>
+<br/>
+
+And PIT report alerts on `equals` and `hashCode` methods.
+
+If these methods are only used from test code, we can replace them with <a href="https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/builder/EqualsBuilder.html#reflectionEquals-java.lang.Object-java.lang.Object-boolean-" target="_blank">`EqualsBuilder.reflectionEquals`</a> from _Apache Commons Lang_:
+
+```
+assertTrue(reflectionEquals(actualObject, expectedObject));
+```
+
+In that case, we can succeed in killing every mutation:
+
+<center>
+<img src="{{site.baseurl}}/assets/img/custom/blog/2017-08-29-coverage/new-pit-statistics.png" alt="PIT Statistics" class="img img-responsive"/>
+</center>
+<br/>
+
+Others prefer <a href="https://projectlombok.org/features/EqualsAndHashCode" target="_blank">Lombok</a> but think about it whether you only need to compare objects.
+
+Regarding _verification_, <a href="https://static.javadoc.io/org.mockito/mockito-core/2.8.47/org/mockito/ArgumentMatchers.html#refEq(T,%20java.lang.String...)">`refEq`</a> is available from <a href="http://site.mockito.org" target="_blank">Mockito</a>.
+
+## Conclusion
```

Comment:
Not sure we need to say that it is the conclusion, but this a matter of style I suppose. So don''t take my word for it; just my liking here :)